New sun models

qualand opened the pull request:
This pull request introduces a new, more robust way to handle sun shape distributions in the simulation codebase. It refactors the previous `DistributionType` usage for sun shapes, replacing it with a dedicated `SunShape` enum, and adds support for additional sun shape models, including the Buie CSR and Limb Darkened model. 

**Sun shape distribution refactor and feature additions:**

* Introduced a new `SunShape` enum to represent sun shape types, replacing the previous use of `DistributionType` for sun sources. This change affects the API for setting and getting sun shapes, and improves clarity and type safety. [[1]](diffhunk://#diff-d0600c276f5815a95ebdb349cf5269c19466f2f02f0cea1139ec09c3ed869315L19-R31) [[2]](diffhunk://#diff-3811459cefdf804c0ddf5818b5ff9c100bd9554b61c873f6a0b82d2e500ffb7fL47-R63)
* Updated the `RaySource` and `Sun` classes to use the new `SunShape` enum, including changes to the `set_shape` and `get_shape` methods and their signatures. [[1]](diffhunk://#diff-d0600c276f5815a95ebdb349cf5269c19466f2f02f0cea1139ec09c3ed869315L35-R44) [[2]](diffhunk://#diff-d0600c276f5815a95ebdb349cf5269c19466f2f02f0cea1139ec09c3ed869315R55-R58) [[3]](diffhunk://#diff-3811459cefdf804c0ddf5818b5ff9c100bd9554b61c873f6a0b82d2e500ffb7fL47-R63)
* Added support for the Buie CSR sun shape model, including validation and calculation of Buie model parameters (`kappa`, `gamma`) in both the `Sun` class and the native runner. [[1]](diffhunk://#diff-639d3b587b75034488b63a9f216477b4a3678244db7efe21c9171d4099e0b419R38-R50) [[2]](diffhunk://#diff-598d94e20a2c2c871fbd4c92acf093f32e324c524809734f3d9d43104c82db52L79-R108)
* Refactored sun shape parsing and initialization in simulation data I/O, adding a dedicated `char_to_sunshape` function and updating sun shape assignment logic. [[1]](diffhunk://#diff-9119309dc173326cca2bdc226d4b15f12c7d83fb7ed0a14103c01c663ade504aR80-R102) [[2]](diffhunk://#diff-9119309dc173326cca2bdc226d4b15f12c7d83fb7ed0a14103c01c663ade504aL269-R300)

**General improvements and code cleanup:**

* Refactored error distribution handling to use `enum class DistributionType`, and updated usages throughout the codebase for improved type safety. [[1]](diffhunk://#diff-bccf5a25379f3e45a534ae2786444c0ba4c45c1a6e414887d4aefb613d86f00cL15-R20) [[2]](diffhunk://#diff-bf04a6fec540c7aea9892fed7488b5ab82149faeeb366856a15d1fa6818cac05L35-R35) [[3]](diffhunk://#diff-9119309dc173326cca2bdc226d4b15f12c7d83fb7ed0a14103c01c663ade504aR65-R68)
* Improved the implementation of surface normal error calculations by removing unused code and simplifying logic. [[1]](diffhunk://#diff-da6f6c418fe88eb3aca3cfaf1d95027e90deea60f5fb9112fd5608870ac48b02R8-L10) [[2]](diffhunk://#diff-da6f6c418fe88eb3aca3cfaf1d95027e90deea60f5fb9112fd5608870ac48b02L30) [[3]](diffhunk://#diff-da6f6c418fe88eb3aca3cfaf1d95027e90deea60f5fb9112fd5608870ac48b02L48-R59) [[4]](diffhunk://#diff-da6f6c418fe88eb3aca3cfaf1d95027e90deea60f5fb9112fd5608870ac48b02L74-L95)

**Simulation runner and data structure updates:**

* Updated simulation runner types and logic to use `SunShape` instead of `DistributionType` for sun shape specification. Added new fields for Buie CSR parameters in simulation data structures. [[1]](diffhunk://#diff-4ca546ed6079a2f775c5907f4d1053c83dfc642295ca18f0e796cc059f4528d3L173-R173) [[2]](diffhunk://#diff-6662ee4e7760660ee41965dd506ee59bb8178b44301792b4c3a7c058e775e6a4L195-R204)
* Updated code to include the new `SunShape` type where relevant, such as in simulation data export and runner initialization.

**Minor corrections and documentation:**

* Added TODO comments and improved documentation in several places to highlight areas for future improvement or implementation. [[1]](diffhunk://#diff-da6f6c418fe88eb3aca3cfaf1d95027e90deea60f5fb9112fd5608870ac48b02R8-L10) [[2]](diffhunk://#diff-9119309dc173326cca2bdc226d4b15f12c7d83fb7ed0a14103c01c663ade504aL269-R300)

qualand committed:
Refactor sun shape handling to use SunShape enum

Introduces a new SunShape enum to distinguish sun shape types from error distribution types, updating all relevant interfaces, implementations, and tests. This improves clarity and maintainability by separating sun shape logic from error distribution logic, and adds support for diffuse and user-defined sun shapes.

qualand committed:
Add Buie CSR and limb-darkened sun shape models

Introduces support for Buie circumsolar ratio (CSR) and limb-darkened sun shape models in the simulation framework. Updates the Sun class and related interfaces to handle the new parameters, implements error checking for CSR, and extends the native runner and error tracing logic to support these sun shapes. Updates unit and regression tests to cover the new models and their error handling.

qualand committed:
Fix branch syntax in CI workflow

qualand committed:
Improve sun shape error handling and comments

Added explanatory comments for unit conversions and sun model behavior. Improved error handling by throwing an exception for unsupported sun shapes. Minor code cleanup and TODO notes for future improvements.

qualand committed:
Merge branch 'new_sun_models' of https://github.com/NREL/SolTrace into new_sun_models

qualand committed:
Limit CI workflow branches to 'develop'

qualand committed:
fixing pull request issues found by Copilot

qualand committed:
Merge branch 'new_sun_models' of https://github.com/NREL/SolTrace into new_sun_models

qualand committed:
Refactor error calculation and update phi computation

Removed unused variables and replaced hardcoded pi value with PI constant in phi calculation for both SurfaceNormalErrors and Errors functions. Added detailed TODO and notes for future refactoring to reduce code duplication and improve error handling.

qualand committed:
Refactor tower demo tests to use shared setup function

Introduced create_tower_demo_simulation_data(bool) to consolidate simulation data setup for all test cases. Updated all tests to use this function, reducing code duplication and improving maintainability. Sun shape modifications for specific tests are now performed after setup. Minor formatting and pointer type consistency improvements.

qualand committed:
Refactor sun shape handling and add new test cases

Removed unnecessary negative angle array filling for user-defined sun shapes in native_runner.cpp and improved switch-case structure. Added sun shape initialization and new test cases for PILLBOX, user-defined, limb darkened, and Buie CSR sun shapes in tower_demo.cpp, including proper simulation parameter setup for error inclusion.

jmaack24 commented on coretrace/simulation_runner/native_runner/native_runner.cpp:
```diff
@@ -76,13 +76,36 @@ namespace SolTrace::NativeRunner
+            this->tsys.Sun.MaxIntensity = 1.0;
+            break;
+        case SunShape::BUIE_CSR: {
+            this->tsys.Sun.MaxAngle = 43.6; // [mrad]
```
Could you move this to a separate function so that the switch statement doesn't get too large?

jmaack24 commented on coretrace/simulation_runner/native_runner/tracing_errors.cpp:
```diff
@@ -160,129 +162,165 @@ void Errors(
-		dist = Sun->ShapeIndex; // sun
 		delop = Sun->Sigma / 1000.0;
+
+		switch (Sun->ShapeIndex)
```
It would be good to move the code from each of the cases of this switch statement to their own functions. The code would be more readable that way.

qualand replied:
opened issue #79 , we can address this when we refactor the errors functions.

jmaack24 reviewed (changes requested):
There were a few places where I thought the code could be made more readable. Otherwise looks good.

qualand committed:
Refactor Buie parameter calculation for sun shape

Moved the Buie (2003) kappa and gamma parameter calculation from native_runner.cpp into a new virtual method calculate_buie_parameters in the RaySource interface, implemented in Sun. This improves code reuse and encapsulation of sun shape logic.
